### ISSUE TITLE
Better separation of concerns in the communication framework

### ISSUE THREAD

**Issue by jngrad** [espressomd/espresso#3907]
**Background**

ESPResSo uses a custom MPI communication framework where everything is managed from the head node. Functions visible in the script interface must be implemented twice: once as a worker function, and once as a head function that contains the same body as the worker function, plus a preamble that triggers the worker function on worker nodes with `mpi_call_*()`. The body usually contains `boost::mpi` and `MPI_*` functions to synchronize data. Often, the body can be implemented only once in a separate free function. These head and worker functions (aka "callbacks") are managed by our `MpiCallbacks` framework which executes them via a visitor pattern.

**Problem statement**

The source file `communication.cpp` is the central place to register callback functions, which should be declared and defined in the relevant cpp resp. hpp files. However for historical reasons, many callbacks have been written directly in `communication.cpp`, which means this file has to know about the implementation details of a variety of classes/structs. This prevents us from hiding implementation details of these classes/structs (e.g. making members private, replacing manual memory management by standard containers, etc.) and can lead to code duplication when the body of the worker function is copy-pasted into the body of the head function. Duplicated code will diverge over time.

**Prior work**

Between 4.1 and 4.3 (waLBerla branch), many callbacks were moved to dedicated source files. For example the NpT callback was moved to `npt.cpp` (d21c7a4e35eb), because `communication.cpp` doesn't need to know the internal structure of the NpT parameters class, it only needs to know that a callback function exists, which is visible via a function prototype in `npt.hpp`. The steepest descent parameters object went one step further and has static storage, making it inaccessible outside the source file:
https://github.com/espressomd/espresso/blob/f760050199b94226b9127b309bfa7db569507605/src/core/integrators/steepest_descent.cpp#L40-L41

**Course of action**

Here are a few candidates that could be simplified further or moved to dedicated files:
* `mpi_bcast_ia_params()` which is written twice, and has now diverged: the worker implementation allocates a pointer with `new` and overwrites the old pointer without a `delete`, leading to a memory leak. Adding `delete` before the `new` doesn't work, because the MPI communication step before that overwrote the address of the allocated memory (see #3906 for details).
* `mpi_bcast_coulomb_params_slave()` needs to know about the Coulomb and dipolar parameter structs. It should rely on `boost::serialization` instead, and have the callbacks written in the relevant electrostatics .cpp files.
* `mpi_gather_stats()` mixes pressure/energy observables with LB communication. The LB communication logic was removed in 4.3. This can be simplified further in 4.3 by replacing the `enum class GatherStats` and the switch statement by a 0/1 integer and a conditional, and removing code duplication between the head and worker functions.

**Challenges**

Moving the remaining callbacks from `communication.cpp` to dedicated .cpp/.hpp files means we have to rely on the `comm_cart` global variable. Passing this global by reference means the dedicated .hpp files that declare callbacks must include MPI headers. However, some parts of the CUDA code are not allowed to include MPI header files, therefore features that are used by CUDA code and rely on MPI communication must either use the `comm_cart` global variable via `extern`, or rely on a second header file, e.g. `npt_mpi.hpp`, to write the prototype of the callback.

**Outlook**

This refactoring will facilitate hiding implementation details of parameter structs and help finding new ways to simplify the callback infrastructure. For example, #3891 removed one level of MPI indirection by having `mpi_steepest_descent()` call `integrate()` via `mpi_steepest_descent_slave()` instead of calling a chain of callbacks. In fact, `mpi_steepest_descent()` could probably call `mpi_integrate_slave()` directly instead of `mpi_steepest_descent_slave()`, removing yet another MPI indirection. This refactoring will also help on the long term if we choose to move in the direction of MPI+OpenMP or pure OpenMP.

**Comment by mkuron** [comment]
> However, some parts of the CUDA code are not allowed to include MPI header files, therefore features that are used by CUDA code and rely on MPI communication must either use the comm_cart global variable via extern, or rely on a second header file, e.g. npt_mpi.hpp, to write the prototype of the callback.

That's an anachronism from the autotools build system (which would call mpic++ and nvcc -- nowadays we use the MPI library search path directly with c++ or nvcc) and from when the OpenMPI headers contained syntax that could not be parsed by nvcc (that was fixed around CUDA 5.5).

**Comment by RudolfWeeber** [comment]
I think many of the affected callbacks should be replaced by script interface code, rather than be refactored as callbacks.

The interaction stuff is a prime candidate, IMO.

The LB stuff in gather_stats will go away with Walberla.

I think the callback registration should happen, where they are implemented as done in lb_walberla_interface.cpp.

**Comment by jngrad** [comment]
Addendum: there are a lot of places in the code where free functions that should have internal linkage end up having external linkage, just so that they could be used in `communication.cpp`. Some of them have a clear warning that they should not be used for any other purpose:
https://github.com/espressomd/espresso/blob/572f85484bcb96d3c2b8f55af8e73791a10be2ca/src/core/particle_data.hpp#L370-L375
But most of them don't. Furthermore, many of them have side effects in the form of events, so accidentally calling one of them can have a noticeable impact on performance. This design is not sustainable. It's also not difficult to fix it, because it's only a matter of moving the callbacks from `communication.cpp` to the .cpp files where they are used: [jngrad/espresso:fix-3906...fix-3907](https://github.com/jngrad/espresso/compare/ef9c69ac9...fix-3907). The free functions can then be made static, or merged into the callbacks directly, so that they are not accessible in other translation units.

**Comment by jngrad** [comment]
Also, there doesn't seem to be a good reason for making the following header files indirectly included in 90 .cpp files:
https://github.com/espressomd/espresso/blob/572f85484bcb96d3c2b8f55af8e73791a10be2ca/src/core/communication.hpp#L55-L57
